garbage_collector.rst

```diff
@@ -0,0 +1,483 @@
+.. _gc:
+
+Design of CPython's Garbage Collector
+=====================================
+
+:Author: Pablo Galindo Salgado
+
+.. highlight:: none
+
+Abstract
+--------
+
+The main garbage collector system of CPython is reference count. The basic idea is
+that CPython counts how many different places there are that have a reference to an
+object. Such a place could be another object, or a global (or static) C variable, or
+a local variable in some C function. When an object’s reference count becomes zero,
+the object is deallocated. If it contains references to other objects, their
+reference count is decremented. Those other objects may be deallocated in turn, if
+this decrement makes their reference count become zero, and so on. The reference
+count field can be examined using the ``sys.getrefcount`` function (notice that the
+value returned by this function is always 1 more as the function also has a reference
+to the object when called):
+
+.. code-block:: python
+
+    >>> x = object()
+    >>> sys.getrefcount(x)
+    2
+    >>> y = x
+    >>> sys.getrefcount(x)
+    3
+    del y
+    >>> sys.getrefcount(x)
+    2
+
+The main problem with the reference count schema is that reference counting
+does not handle reference cycles. For instance, consider this code:
+
+.. code-block:: python
+
+    >>> container = []
+    >>> container.append(container)
+    >>> sys.getrefcount(container)
+    3
+    >>> del container
+
+In this example, ``container`` holds a reference to itself, so even when we remove
+our reference to it (the variable "container") the reference count never falls to 0
+because it still has its own internal reference and therefore it will never be
+cleaned just by simple reference counting. For this reason some additional machinery
+is needed to clean these reference cycles between objects once they become
+unreachable. We normally refer to this additional machinery as the Garbage Collector,
+but technically reference counting is also a form of garbage collection.
```

Comment:
Comments in the code use "cyclic GC" (and variations of that). Can we use that term when we need to contrast "the GC" with refcounting? For example:
```
... For this reason some additional machinery
is needed to clean these reference cycles between objects once they become
unreachable. This is the cyclic garbage collector, usually called just
Garbage Collector (GC), even though reference counting is also a form
of garbage collection.
```

Reply:
Thanks! Done in 4321593